servicetalk-http-api/src/main/java/io/servicetalk/http/api/StreamingHttpPayloadHolder.java

```diff
@@ -17,127 +17,211 @@
-        if (payloadBody != null) {
-            splitTrailersIfRequired();
+    @Deprecated
+    void transformRawPayloadBody(UnaryOperator<Publisher<?>> transformer) {
+        payloadInfo.setOnlyEmitsBuffer(false); // We lose type information and cannot guarantee emitting buffers.
+        payloadInfo.setMayHaveTrailers(true); // We don't know if the user will insert trailers or not, assume they may.
+        if (messageBody != null && payloadInfo.mayHaveTrailers()) {
+            final Publisher<?> oldMessageBody = messageBody;
+            messageBody = defer(() -> {
+                Processor<HttpHeaders, HttpHeaders> trailersProcessor = newSingleProcessor();
+                return merge(transformer.apply(oldMessageBody.liftSync(
+                        new PreserveTrailersOperator(trailersProcessor))), fromSource(trailersProcessor))
+                        .subscribeShareContext();
+            });
+        } else {
+            // It is possible the resulting transformed publisher contains a trailers object, but this is not supported
+            // by this API. Instead use the designated trailers transform methods.
+            messageBody = transformer.apply(messageBody());
         }
-        payloadBody = transformer.apply(emptyOrRawPayload());
     }
 
-    public <T> void transform(final TrailersTransformer<T, Buffer> trailersTransformer) {
-        transformWithTrailersUnchecked(false, trailersTransformer);
+    void transformMessageBody(UnaryOperator<Publisher<?>> transformer) {
+        payloadInfo.setOnlyEmitsBuffer(false); // We lose type information and cannot guarantee emitting buffers.
+        // The transformation does not support changing the presence of trailers in the message body.
+        messageBody = transformer.apply(messageBody());
+    }
+
+    <T> void transform(final TrailersTransformer<T, Buffer> trailersTransformer) {
+        transformWithTrailersUnchecked(trailersTransformer, o -> (Buffer) o);
     }
 
-    public <T> void transformRaw(final TrailersTransformer<T, Object> trailersTransformer) {
-        transformWithTrailersUnchecked(true, trailersTransformer);
+    @Deprecated
+    <T> void transformRaw(final TrailersTransformer<T, Object> trailersTransformer) {
+        payloadInfo.setOnlyEmitsBuffer(false); // We lose type information and cannot guarantee emitting buffers.
+        transformWithTrailersUnchecked(trailersTransformer, identity());
     }
 
     Single<PayloadAndTrailers> aggregate() {
         payloadInfo.setSafeToAggregate(true);
-        return aggregatePayloadAndTrailers(payloadBodyAndTrailers(), allocator);
+        return aggregatePayloadAndTrailers(messageBody(), allocator);
+    }
+
+    private <T, P> void transformWithTrailersUnchecked(final TrailersTransformer<T, P> trailersTransformer,
+                                                       final Function<Object, P> caster) {
+        if (version == HTTP_1_1) {
```

Comment:
Consider using a check similar to what you have in `BlockingStreamingToStreamingService`:
```
metaData.version().major() == 1 && metaData.version().minor() > 0
```

Every time I see this, thinking about implementing `Comparable` for `HttpProtocolVersion` :)

Reply:
good point. I will update `BlockingStreamingToStreamingService` to use `== HTTP_1_1 ` which is simpler and 1.2+ isn't a thing.

Reply:
Ok, both ways are fine. HTTP/1.2 is not official, I was just thinking that someone can define it (we allow custom `HttpProtocolVersion`) with some custom subset of headers or something like that.